command/command.go

```diff
@@ -9,121 +9,154 @@ import (
 	"strings"
 )
 
-// ----------
-
-// Model ...
-type Model struct {
+// Command ...
+type Command interface {
+	SetDir(dir string) Command
+	SetStdout(stdout io.Writer) Command
+	SetStderr(stdout io.Writer) Command
+	SetStdoutAndStderr() Command
+	SetEnvs(envs ...string) Command
+	AppendEnvs(envs ...string) Command
+	Args() []string
+	PrintableCommandArgs() string
+	Run() error
+	RunAndReturnTrimmedOutput() (string, error)
+	RunAndReturnExitCode() (int, error)
+	RunAndReturnTrimmedCombinedOutput() (string, error)
+}
+
+// Factory ...
+type Factory func(name string, args ...string) Command
+
+type cmdWrapper struct {
 	cmd *exec.Cmd
 }
 
-// New ...
-func New(name string, args ...string) *Model {
-	return &Model{
+func newCmdWrapper(name string, args ...string) *cmdWrapper {
+	return &cmdWrapper{
 		cmd: exec.Command(name, args...),
 	}
 }
 
-// NewWithStandardOuts - same as NewCommand, but sets the command's
+// New ...
+func New(name string, args ...string) Command {
+	return newCmdWrapper(name, args...)
+}
+
+// NewWithStandardOuts - same as New, but sets the command's
 // stdout and stderr to the standard (OS) out (os.Stdout) and err (os.Stderr)
-func NewWithStandardOuts(name string, args ...string) *Model {
-	return New(name, args...).SetStdout(os.Stdout).SetStderr(os.Stderr)
+// Deprecated: Use New and SetStdoutAndStderr instead.
+func NewWithStandardOuts(name string, args ...string) Command {
+	return newCmdWrapper(name, args...).SetStdout(os.Stdout).SetStderr(os.Stderr)
 }
 
 // NewWithParams ...
-func NewWithParams(params ...string) (*Model, error) {
+// Deprecated.
+func NewWithParams(params ...string) (Command, error) {
 	if len(params) == 0 {
 		return nil, errors.New("no command provided")
 	} else if len(params) == 1 {
-		return New(params[0]), nil
+		return newCmdWrapper(params[0]), nil
 	}
 
-	return New(params[0], params[1:]...), nil
+	return newCmdWrapper(params[0], params[1:]...), nil
 }
 
 // NewFromSlice ...
-func NewFromSlice(slice []string) (*Model, error) {
+// Deprecated.
```

Comment:
I'd just delete the "deprecated" stuff. We'll update the usage sites when necessary. Other than that it is vendored

Reply:
I'm okay with both approaches, but I can imagine with this many deprecations it would be overwhelming for some external repositories to migrate away from all methods/functions that would be deleted by this.

We can delete them, but I'd resort to deprecations for now. Those are correctly highlighted by IDEs as well, so they hardly go under the radar.